modules/reitit-ring/src/reitit/ring.cljc

```diff
@@ -65,6 +65,11 @@
       ([request respond _]
        (respond (handle request))))))
 
+(def ring-endpoint (r/mk-intermediate-endpoint-transform [{:kss (for [method http-methods] [method :handler])
+                                                           :transform :inherit}
+                                                          {:kss       [[:name]]
+                                                           :transform :consume}]))
+
```

Comment:
Contract for `mk-intermediate-endpoint-transform` is not defined and looks bit complex. Ring already has a custom `:compile` that knows how to handle the http-methods. Would be great if that could be used instead. Does the `:compile` happen too late so it can't be used here? If it could, all the routers would work the same way and we could just declare the `:endpoint` options as data, instead of a function callback.

Reply:
also, there are comments that the `:name` [should be allowed only on leafs](https://github.com/metosin/reitit/issues/271#issuecomment-546647754). If that would be implemented, it would make this much simpler?

Reply:
> Contract for `mk-intermediate-endpoint-transform` is not documented and looks bit complex. 

`mk-intermediate-endpoint-transform` has a docstring which tells how it works. In any case, it is an implementation detail, which should be out of user's eyes. I agree that it seems complex, but I was struggling how to express (tersely) what kind of transform will be performed on different key seqs of route data. For reference, this implements the idea, that @ikitommi suggested, that some route data would be consumed by the intermediate route, while other would be passed down to child routes. 

The actual implementation of the endpoint transformation (which `mk-intermediate-endpoint-transform` creates for the new default endpoint resolution) is documented here: https://github.com/metosin/reitit/pull/326/files#diff-7d84e60643070c30a98ce39e202d308bR426

> reitit-ring already has a custom `:compile` that knows how to handle the http-methods. Would be great if that could be used instead. Does the `:compile` happen too late so it can't be used here? 

Currently, the intermediate routes do not survive the first step of endpoint (route) resolution, so the compile is too late with current behaviour. I'm pretty sure I wrote that somewhere, but I'm not able to find it. 

> If it could, all the routers would work the same way and we could just declare the `:endpoint` options as data, instead of a function callback.

Hmm. This is kind of what I was aiming for... Would it be better to have the "endpoint transform" as completely internal to route resolution, and have router-specific configuration as just data: ?
```
(def ring-endpoint [{:kss (for [method http-methods] [method :handler])
                               :transform :inherit}
                    {:kss       [[:name]]
                     :transform :consume}]))
```

Reply:
> also, there are comments that the :name should be allowed only on leafs. If that would be implemented, it would make this much simpler?

I can't see where this is mentioned in the conversation? 

This (allowing `:name` only on leaves) would defeat the purpose of this whole PR, which is to enable intermediate routes. The rule that is applied in this PR is that paths with `:name` will likely be referenced by the user, so they can act as routes (and matched by name). 

However, the behaviour implemented in this PR (that you @ikitommi suggested) does seem to reduce conflicts with `:name`, as the default behaviour would now be to consume the name at the path. Currently, it is passed it to all children, which leads to duplicate names for routes. 

Maybe we could talk this through, so we're on the same map on what should be accomplished here.

Reply:
Also, I would be more than happy for simpler, but concise, way of defining which paths are routes (predicate "endpoint?") while specifying the transform of route data (`consume` vs. `inherit`). :sweat_smile:

Reply:
@tonivanhala I believe @ikitommi is referencing https://github.com/metosin/reitit/issues/271 when mentioning the disucssion around `:name`. I do agree that the issues discussed in there and this PR are somewhat contradictory.

One potential way to bring them back in sync might be to assume a `:method :get` configuration for an intermediate that has a `:name` and doesn't specify the `:method` itself. I believe that if that were the case, all of the changes mentioned in #271 could still be implemented on top of this PR.

I have run into the confusing behavior described and resolved by this PR so I definitely see value in it. @ikitommi thoughts on if the behavior in this PR is desired given your comments in #271 ?